File: mobile/lib/logic/score.dart
import '../models/phone.dart';

class _Range {
  const _Range(this.min, this.max);
  final double min;
  final double max;
}

class ScoreEngine {
  ScoreEngine(this.phones) : _catalog = _Catalog.from(phones);

  final List<Phone> phones;
  final _Catalog _catalog;

  double technical(Phone phone) {
    final performance = _norm(phone.benchmarks.antutu.toDouble(), _catalog.antutu) * 0.65 +
        _norm(phone.benchmarks.geekbenchMulti.toDouble(), _catalog.geek) * 0.35;
    final camera = _norm(_cameraRaw(phone), _catalog.camera);
    final display = _norm(_displayRaw(phone), _catalog.display);
    final battery = _norm(_batteryRaw(phone), _catalog.battery);
    final memory = _norm(_memoryRaw(phone), _catalog.memory);
    final body = _norm(_bodyRaw(phone), _catalog.body);
    final extras = _featureRaw(phone);
    final total = performance * 0.22 +
        camera * 0.18 +
        display * 0.14 +
        battery * 0.16 +
        memory * 0.1 +
        body * 0.08 +
        extras * 0.12;
    return _round1(total);
  }

  Map<String, double> categories(Phone phone) => {
        'performans': _round1(_norm(phone.benchmarks.antutu.toDouble(), _catalog.antutu)),
        'kamera': _round1(_norm(_cameraRaw(phone), _catalog.camera)),
        'ekran': _round1(_norm(_displayRaw(phone), _catalog.display)),
        'batarya': _round1(_norm(_batteryRaw(phone), _catalog.battery)),
        'bellek': _round1(_norm(_memoryRaw(phone), _catalog.memory)),
      };

  double recency(Phone phone) {
    final age = DateTime.now().difference(phone.releaseDate).inDays.toDouble();
    return _clamp(100 - age / 8);
  }

  double smart(Phone phone, double userAverage) =>
      phone.popularity * 0.38 +
      technical(phone) * 0.32 +
      userAverage * 10 * 0.18 +
      recency(phone) * 0.12;

  double index(double technicalScore, double userAverage) =>
      _round1(technicalScore * 0.55 + userAverage * 10 * 0.45);
}

class _Catalog {
  _Catalog({
    required this.antutu,
    required this.geek,
    required this.camera,
    required this.display,
    required this.battery,
    required this.memory,
    required this.body,
  });

  final _Range antutu;
  final _Range geek;
  final _Range camera;
  final _Range display;
  final _Range battery;
  final _Range memory;
  final _Range body;

  factory _Catalog.from(List<Phone> phones) {
    final scored = phones.where((phone) => phone.image.isNotEmpty).toList();
    final base = scored.isNotEmpty ? scored : phones;
    return _Catalog(
      antutu: _range(base.map((p) => p.benchmarks.antutu.toDouble())),
      geek: _range(base.map((p) => p.benchmarks.geekbenchMulti.toDouble())),
      camera: _range(base.map(_cameraRaw)),
      display: _range(base.map(_displayRaw)),
      battery: _range(base.map(_batteryRaw)),
      memory: _range(base.map(_memoryRaw)),
      body: _range(base.map(_bodyRaw)),
    );
  }
}

double _cameraRaw(Phone p) {
  final rear = p.camera.rear.fold<double>(0, (s, l) => s + l.mp);
  final dxo = p.camera.dxomark ?? (rear > 0 ? 118 : 0);
  return dxo * 2 + rear;
}

double _displayRaw(Phone p) =>
    p.display.size * 8 + p.display.refreshRate * 0.25 + p.display.brightness / 80 + p.display.ppi / 12;

double _batteryRaw(Phone p) {
  final capacity = p.battery.capacity ?? p.battery.estimatedHours * 155;
  return capacity + p.battery.wiredWatt * 8 + p.battery.estimatedHours * 40;
}

double _memoryRaw(Phone p) => p.memory.ram * 18 + p.memory.storage * 0.08;

double _bodyRaw(Phone p) {
  final weight = p.body.weight == 0 ? 200 : p.body.weight;
  final thickness = p.body.thickness == 0 ? 8.2 : p.body.thickness;
  return _ipRank(p.body.ipRating) * 2 + (280 - weight) + (12 - thickness) * 8;
}

double _featureRaw(Phone p) {
  var score = 40.0;
  if (p.connectivity.network5g) score += 10;
  if (p.connectivity.nfc) score += 6;
  if (p.connectivity.esim) score += 5;
  if (p.battery.wirelessWatt > 0) score += 8;
  if (p.battery.reverse) score += 4;
  if (p.audio.dolby) score += 4;
  if (p.features.stylus) score += 6;
  if (p.features.faceUnlock) score += 4;
  if (p.display.alwaysOn) score += 4;
  if (p.display.hdr) score += 4;
  if (p.connectivity.wifi.contains('7')) score += 5;
  return _clamp(score);
}

int _ipRank(String rating) {
  final match = RegExp(r'IP(\d)(\d)', caseSensitive: false).firstMatch(rating);
  if (match == null) return 0;
  return int.parse(match.group(1)!) * 10 + int.parse(match.group(2)!);
}

_Range _range(Iterable<double> values) {
  final list = values.toList();
  return _Range(list.reduce((a, b) => a < b ? a : b), list.reduce((a, b) => a > b ? a : b));
}

double _norm(double value, _Range range) {
  if (range.max == range.min) return 70;
  return _clamp(((value - range.min) / (range.max - range.min)) * 100);
}

double _clamp(double value) => value.clamp(0, 100);

double _round1(double value) => (value * 10).round() / 10;

File: src/lib/score.ts
import type { Phone } from "@/types/phone";
import { phones } from "@/data/phones";

function clamp(value: number, min = 0, max = 100) {
  return Math.min(max, Math.max(min, value));
}

function rangeOf(values: number[]) {
  const finite = values.filter((value) => Number.isFinite(value));
  if (finite.length === 0) return { min: 0, max: 1 };
  const min = Math.min(...finite);
  const max = Math.max(...finite);
  return { min, max };
}

function norm(value: number, min: number, max: number, invert = false) {
  if (max === min) return 70;
  const raw = ((value - min) / (max - min)) * 100;
  return clamp(invert ? 100 - raw : raw);
}

function ipRank(rating: string) {
  const match = rating.match(/IP(\d)(\d)/i);
  if (!match) return 0;
  return Number(match[1]) * 10 + Number(match[2]);
}

function rearMp(phone: Phone) {
  return phone.camera.rear.reduce((sum, lens) => sum + lens.mp, 0);
}

function cameraRaw(phone: Phone) {
  const dxo = phone.camera.dxomark ?? (rearMp(phone) > 0 ? 118 : 0);
  return dxo * 2 + rearMp(phone);
}

const scored = phones.filter((phone) => phone.image);
const catalogBase = scored.length ? scored : phones;

const catalog = {
  antutu: rangeOf(catalogBase.map((p) => p.benchmarks.antutu)),
  geek: rangeOf(catalogBase.map((p) => p.benchmarks.geekbenchMulti)),
  camera: rangeOf(catalogBase.map((p) => cameraRaw(p))),
  display: rangeOf(catalogBase.map((p) => displayRaw(p))),
  battery: rangeOf(catalogBase.map((p) => batteryRaw(p))),
  memory: rangeOf(catalogBase.map((p) => memoryRaw(p))),
  body: rangeOf(catalogBase.map((p) => bodyRaw(p))),
};

function displayRaw(phone: Phone) {
  return (
    phone.display.size * 8 +
    phone.display.refreshRate * 0.25 +
    phone.display.brightness / 80 +
    phone.display.ppi / 12
  );
}

function batteryRaw(phone: Phone) {
  const capacity = phone.battery.capacity ?? phone.battery.estimatedHours * 155;
  return capacity + phone.battery.wiredWatt * 8 + phone.battery.estimatedHours * 40;
}

function memoryRaw(phone: Phone) {
  return phone.memory.ram * 18 + phone.memory.storage * 0.08;
}

function bodyRaw(phone: Phone) {
  const weight = phone.body.weight || 200;
  const thickness = phone.body.thickness || 8.2;
  return ipRank(phone.body.ipRating) * 2 + (280 - weight) + (12 - thickness) * 8;
}

function featureRaw(phone: Phone) {
  let score = 40;
  if (phone.connectivity.network5g) score += 10;
  if (phone.connectivity.nfc) score += 6;
  if (phone.connectivity.esim) score += 5;
  if (phone.battery.wirelessWatt > 0) score += 8;
  if (phone.battery.reverse) score += 4;
  if (phone.audio.dolby) score += 4;
  if (phone.features.stylus) score += 6;
  if (phone.features.faceUnlock) score += 4;
  if (phone.display.alwaysOn) score += 4;
  if (phone.display.hdr) score += 4;
  if (phone.connectivity.wifi.includes("7")) score += 5;
  return clamp(score);
}

export function technicalScore(phone: Phone) {
  const performance =
    norm(phone.benchmarks.antutu, catalog.antutu.min, catalog.antutu.max) * 0.65 +
    norm(phone.benchmarks.geekbenchMulti, catalog.geek.min, catalog.geek.max) * 0.35;

  const camera = norm(cameraRaw(phone), catalog.camera.min, catalog.camera.max);

  const display = norm(displayRaw(phone), catalog.display.min, catalog.display.max);
  const battery = norm(batteryRaw(phone), catalog.battery.min, catalog.battery.max);
  const memory = norm(memoryRaw(phone), catalog.memory.min, catalog.memory.max);
  const body = norm(bodyRaw(phone), catalog.body.min, catalog.body.max);
  const extras = featureRaw(phone);

  const total =
    performance * 0.22 +
    camera * 0.18 +
    display * 0.14 +
    battery * 0.16 +
    memory * 0.1 +
    body * 0.08 +
    extras * 0.12;

  return Math.round(total * 10) / 10;
}

export function categoryScores(phone: Phone) {
  return {
    performans: Math.round(norm(phone.benchmarks.antutu, catalog.antutu.min, catalog.antutu.max) * 10) / 10,
    kamera: Math.round(norm(cameraRaw(phone), catalog.camera.min, catalog.camera.max) * 10) / 10,
    ekran: Math.round(norm(displayRaw(phone), catalog.display.min, catalog.display.max) * 10) / 10,
    batarya: Math.round(norm(batteryRaw(phone), catalog.battery.min, catalog.battery.max) * 10) / 10,
    bellek: Math.round(norm(memoryRaw(phone), catalog.memory.min, catalog.memory.max) * 10) / 10,
  };
}

export function recencyScore(phone: Phone) {
  const age = (Date.now() - new Date(phone.releaseDate).getTime()) / (1000 * 60 * 60 * 24);
  return clamp(100 - age / 8);
}

export function smartScore(phone: Phone, userAverage: number) {
  return (
    phone.popularity * 0.38 +
    technicalScore(phone) * 0.32 +
    userAverage * 10 * 0.18 +
    recencyScore(phone) * 0.12
  );
}

export function globalIndex(technical: number, userAverage: number) {
  return Math.round((technical * 0.55 + userAverage * 10 * 0.45) * 10) / 10;
}

export function scoreTone(score: number) {
  if (score >= 88) return "high";
  if (score >= 76) return "mid";
  return "low";
}
